openslides_backend/action/speaker/end_speach.py

```diff
@@ -0,0 +1,51 @@
+import time
+from typing import Iterable
+
+from ...models.models import Speaker
+from ...shared.exceptions import ActionException
+from ...shared.interfaces import Event
+from ...shared.patterns import FullQualifiedId
+from ..base import Action, ActionPayload, DataSet, WriteRequestElement
+from ..default_schema import DefaultSchema
+from ..register import register_action
+
+
+@register_action("speaker.end_speach")
+class SpeakerEndSpeach(Action):
+    """
+    Action to stop speakers.
+    """
+
+    model = Speaker()
+    schema = DefaultSchema(Speaker()).get_default_schema(
+        required_properties=["id"],
+        title="End speach schema",
+        description="Schema to stop a speaker's speach.",
+    )
+
+    def prepare_dataset(self, payload: ActionPayload) -> DataSet:
+        data = []
+        for instance in payload:
+            speaker = self.fetch_model(
+                FullQualifiedId(self.model.collection, instance["id"]),
+                mapped_fields=["begin_time", "end_time"],
+            )
+            if speaker.get("begin_time") is None or speaker.get("end_time") is not None:
+                raise ActionException(
+                    f"Speaker {instance['id']} is not speaking at the moment."
+                )
+            instance["end_time"] = round(time.time())
+            data.append({"instance": instance})
+        return {"data": data}
+
+    def create_write_request_elements(
+        self, dataset: DataSet
+    ) -> Iterable[WriteRequestElement]:
```

Comment:
If you (rebase and) subclass from `UpdateAction`, you can use the `get_updated_models` function instead of `prepare_dataset` and don't have to overwrite this function. This should simplify your code by a good amount.

Reply:
Good idea. I also change the return value from List to Iterable so I can use yield statement.